### PR TITLE
Refactor AlicaTime and AlicaClock

### PR DESCRIPTION
**Pull request by abolinsky**
# **DO NOT REVIEW UNTIL hs_constify_engine_model HAS BEEN MERGED!!!**

**AlicaClock** was formally three classes: **IAlicaClock**, **AlicaSystemClock**, and **AlicaROSClock**.
The two base classes were implemented identically under the hood, and there was no need for virtual inheritance, so all three classes were combined into **AlicaClock**.

All instances of time durations or time, which were formally combinations of unsigned long long, unsigned long, long, int, and double, have been consolidated into a single class called **AlicaTime**. All instances of messy magic numbers and comments denoting units of time have been replaced by calling member functions within **AlicaTime**.

### CODE REVIEW

File: constraintsolver/include/GSolver.h [hunk 1]
```diff
@@ -9,16 +9,18 @@
...
 #include <AutoDiff.h>
+#include <engine/AlicaClock.h>
 
 #include <memory>
 #include <vector>
 
 #include <fstream>
 
 using namespace std;
+using alica::AlicaTime;
```
Comment (gauthamm) [inline constraintsolver/include/GSolver.h]: not needed, you are referencing to AlicaTime and AlicaClock only in the namespace alica anyway

File: world_model/include/supplementary/WorldModel.h [hunk 2]
```diff
@@ -1,6 +1,8 @@
 #pragma once
 
-#include <supplementary/InfoBuffer.h>
+#include <engine/AlicaClock.h>
+
+using alica::AlicaTime;
```
Comment (gauthamm) [inline world_model/include/supplementary/WorldModel.h]: not needed

File: constraintsolver/include/CNSat.h [hunk 3]
```diff
@@ -9,14 +9,14 @@
 #define CNSAT_H_
 
 #include "types/Assignment.h"
+#include "engine/AlicaClock.h"
```
Comment (hendrik-skubch) [inline constraintsolver/include/CNSat.h]: Why do you remove the forward declaration here?

File: event_handling/include/supplementary/NotifyTimer.h [hunk 4]
```diff
@@ -1,13 +1,16 @@
 #pragma once
 
 #include "ITrigger.h"
+#include "engine/AlicaClock.h"
```
Comment (hendrik-skubch) [inline event_handling/include/supplementary/NotifyTimer.h]: This is not so cool. While the change looks nice you have a circular dependency between engine and event handling now. This class should not depend on alica-engine.

**Review by hendrik-skubch — CHANGES_REQUESTED** [review]

File: constraintsolver/include/CNSat.h (hunk 3 above)
Comment (abolinsky) [reply constraintsolver/include/CNSat.h]: Ah, I overlooked the pointer to AlicaClock. In the case of pointers to objects, the class can just be forward declared. Fixed.

**Review by hendrik-skubch — APPROVED** [review]

**Pull request merged**